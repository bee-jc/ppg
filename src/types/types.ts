export interface Episode {
  id: number;
  name: string;
  image: { original: string; medium?: string };
  summary: string;
  season: number;
  imageUrl?: string;
}

export interface Season {
  id: number;
  url: string;
  number: number;
  name: string;
  episodeOrder: number;
  premiereDate: string | null;
  endDate: string | null;
  network: {
    id: number;
    name: string;
    country: {
      name: string;
      code: string;
      timezone: string;
    };
  };
  webChannel: any;
  image: {
    original: string | null;
  } | null;
  summary: string | null;
}

export interface Show {
  id: number;
  name: string;
  image: { original: string };
  summary: string;
}

export interface EpisodeState {
  details: Episode;
  loading: boolean;
  error: string | null;
}

export interface ShowState {
  details: Show;
  episodes: Episode[];
  seasons: Season[];
  loading: boolean;
}

export interface SeasonWithEpisodes extends Season {
  episodes: Episode[];
}
