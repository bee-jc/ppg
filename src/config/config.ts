const API_BASE_URL = 'https://api.tvmaze.com';

export const API_URLS = {
  fetchShowDetails: (id: number) => `${API_BASE_URL}/shows/${id}`,
  fetchEpisodes: (id: number) => `${API_BASE_URL}/shows/${id}/episodes`,
  fetchEpisodeDetails: (id: number) => `${API_BASE_URL}/episodes/${id}`,
  fetchSeasons: (id: number) => `${API_BASE_URL}/shows/${id}/seasons`,
};

// Configuration for Redux slice
export const SLICE_NAMES = {
  show: 'show',
  episode: 'episode',
};

// Redux action types
export const ACTION_TYPES = {
  getEpisodeDetails: `${SLICE_NAMES.episode}/getEpisodeDetails`,
  getEpisodes: `${SLICE_NAMES.show}/getEpisodes`,
  getSeasons: `${SLICE_NAMES.show}/getSeasons`,
  getShowDetails: `${SLICE_NAMES.show}/getShowDetails`,
};

// Error Messages
export const ERROR_MESSAGES = {
  fetchEpisodeDetailsError: 'Failed to fetch episode details',
};

export const POWERPUFF_GIRLS_SHOW_ID = 1955;

export const HEADLINE_NOTE =
  'The Powerpuff Girls is an award-winning animated action series about three super-powered little girls Bubbles, Blossom and Buttercup with one very big mission: saving the world before bedtime.';
