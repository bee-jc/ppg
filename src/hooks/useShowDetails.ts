import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RootState, AppDispatch } from '../redux/store';
import {
  getShowDetails,
  getEpisodes,
  getSeasons,
} from '../redux/slices/showSlice';
import { groupEpisodesBySeason } from '../utils/groupEpisodesBySeason';

export const useShowDetails = (showId: number) => {
  const dispatch = useDispatch<AppDispatch>();
  const { details, episodes, seasons } = useSelector(
    (state: RootState) => state.show
  );

  const [activeSeasonId, setActiveSeasonId] = useState<number | null>(null);

  useEffect(() => {
    dispatch(getShowDetails(showId));
    dispatch(getEpisodes(showId));
    dispatch(getSeasons(showId));
  }, [dispatch, showId]);

  const seasonsWithEpisodes = groupEpisodesBySeason(episodes, seasons);

  useEffect(() => {
    if (seasonsWithEpisodes.length > 0 && activeSeasonId === null) {
      setActiveSeasonId(seasonsWithEpisodes[0].id);
    }
  }, [seasonsWithEpisodes, activeSeasonId]);

  return {
    details,
    seasonsWithEpisodes,
    activeSeasonId,
    setActiveSeasonId,
  };
};
