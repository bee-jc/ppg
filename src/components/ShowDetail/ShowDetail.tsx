import React from 'react';

import { useShowDetails } from '../../hooks/useShowDetails';

import SeasonDetails from './components/SeasonDetail/SeasonDetails';
import SeasonEpisodes from './components/SeasonEpisodes/SeasonEpisodes';
import { POWERPUFF_GIRLS_SHOW_ID, HEADLINE_NOTE } from '../../config/config';

import {
  ShowDetailWrapper,
  ShowDetailMain,
  ShowHero,
  ShowContent,
  SeasonTabs,
  SeasonTab,
  EpisodesWrapper,
  ShowImage,
  ShowDescription,
} from '../../styles';

export const ShowDetail: React.FC = () => {
  const { details, seasonsWithEpisodes, activeSeasonId, setActiveSeasonId } =
    useShowDetails(POWERPUFF_GIRLS_SHOW_ID);

  if (!seasonsWithEpisodes.length) {
    return <div>Loading...</div>;
  }

  return (
    <ShowDetailWrapper>
      <ShowDetailMain>
        <ShowHero backgroundImage={details.image?.original}>
          <ShowImage
            src={details.image?.original || 'placeholder-image-url'}
            alt="Show Image"
          />
          <ShowDescription>{HEADLINE_NOTE}</ShowDescription>
        </ShowHero>
        <ShowContent>
          <SeasonTabs>
            {seasonsWithEpisodes.map((season) => (
              <SeasonTab
                key={season.id}
                isActive={season.id === activeSeasonId}
                onClick={() => setActiveSeasonId(season.id)}
              >
                <SeasonDetails
                  seasonNumber={season.number}
                  episodeCount={season.episodes.length}
                />
              </SeasonTab>
            ))}
          </SeasonTabs>
          <EpisodesWrapper>
            {seasonsWithEpisodes
              .filter((season) => season.id === activeSeasonId)
              .map((season) => (
                <SeasonEpisodes key={season.id} season={season} />
              ))}
          </EpisodesWrapper>
        </ShowContent>
      </ShowDetailMain>
    </ShowDetailWrapper>
  );
};
