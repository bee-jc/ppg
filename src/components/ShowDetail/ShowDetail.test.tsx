import { render, screen, fireEvent } from '@testing-library/react';
import { ShowDetail } from './ShowDetail';
import { useShowDetails } from '../../hooks/useShowDetails';
import { POWERPUFF_GIRLS_SHOW_ID } from '../../config/config';
import { SeasonWithEpisodes, Episode } from '../../types/types';
import { MemoryRouter } from 'react-router-dom';

// Mock the useShowDetails hook
jest.mock('../../hooks/useShowDetails', () => ({
  useShowDetails: jest.fn(),
}));

// Mock data
const mockDetails = {
  id: POWERPUFF_GIRLS_SHOW_ID,
  name: 'Powerpuff Girls',
  image: { original: 'http://example.com/show.jpg' },
  summary: 'This is a test summary.',
};

const mockSeasonsWithEpisodes: SeasonWithEpisodes[] = [
  {
    id: 1,
    number: 1,
    name: 'Season 1',
    episodeOrder: 2,
    premiereDate: '2023-01-01',
    endDate: '2023-01-02',
    network: {
      id: 1,
      name: 'Network',
      country: {
        name: 'Country',
        code: 'CT',
        timezone: 'Timezone',
      },
    },
    webChannel: null,
    image: { original: 'http://example.com/season.jpg' },
    summary: 'Summary for Season 1',
    episodes: [
      {
        id: 1,
        name: 'Episode 1',
        image: {
          original: 'http://example.com/image1.jpg',
          medium: 'http://example.com/medium1.jpg',
        },
        summary: 'Summary for Episode 1',
        season: 1,
      },
    ],
    url: 'http://example.com/season-url',
  },
];

const mockActiveSeasonId = 1;
const mockSetActiveSeasonId = jest.fn();

describe('ShowDetail', () => {
  beforeEach(() => {
    (useShowDetails as jest.Mock).mockReturnValue({
      details: mockDetails,
      seasonsWithEpisodes: mockSeasonsWithEpisodes,
      activeSeasonId: mockActiveSeasonId,
      setActiveSeasonId: mockSetActiveSeasonId,
    });
  });

  it('should render correctly with mock data', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ShowDetail />
      </MemoryRouter>
    );

    // Check if season tabs are rendered
    expect(screen.getByText('Season 1')).toBeInTheDocument();

    // Check if episode list is rendered
    expect(screen.getByText('Episode 1')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call setActiveSeasonId when a season tab is clicked', () => {
    render(
      <MemoryRouter>
        <ShowDetail />
      </MemoryRouter>
    );

    const seasonTab = screen.getByText('S01');
    expect(seasonTab).toBeInTheDocument();

    // Click on the season tab
    fireEvent.click(seasonTab);

    // Check if setActiveSeasonId was called
    expect(mockSetActiveSeasonId).toHaveBeenCalledWith(1);
  });
});
