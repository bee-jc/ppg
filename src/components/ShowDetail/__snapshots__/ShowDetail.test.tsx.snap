// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetail should render correctly with mock data 1`] = `
<DocumentFragment>
  <div
    class="css-r4r7y2"
  >
    <main
      class="css-1iv0tiv"
    >
      <section
        class="css-ec5vh9"
      >
        <img
          alt="Show Image"
          class="css-4ofcwm"
          src="http://example.com/show.jpg"
        />
        <div
          class="css-1c5ad3y"
        >
          The Powerpuff Girls is an award-winning animated action series about three super-powered little girls Bubbles, Blossom and Buttercup with one very big mission: saving the world before bedtime.
        </div>
      </section>
      <div
        class="css-oauwje"
      >
        <div
          class="css-1p822b5"
        >
          <button
            class="css-13nqquo"
          >
            <div
              class="css-1vdtga9"
            >
              S01
            </div>
            <div
              class="css-1wqfxc5"
            >
              (1 episodes)
            </div>
          </button>
        </div>
        <div
          class="css-1hc9joy"
        >
          <div
            class="css-i444qj"
          >
            Season 1
          </div>
          <div
            class="css-1fagh72"
          >
            <a
              class="css-3tqedt"
              href="/episode/1"
            >
              <div
                class="css-1v8ruy6"
              >
                <img
                  alt="Episode 1"
                  class="css-5y3q7n"
                  src="http://example.com/medium1.jpg"
                />
              </div>
              <div
                class="css-14mp7s"
              >
                <div
                  class="css-tqpnt9"
                >
                  Episode 1
                </div>
                <p
                  class="css-juyb0m"
                >
                  Summary for Episode 1
                </p>
              </div>
            </a>
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
